File: internal/tui/selection/range.go
package selection

import "strings"

const minCopyRunes = 3

type Point struct {
	Line   int
	Column int
}

type Range struct {
	Anchor Point
	Focus  Point
	Active bool
}

func (r Range) Normalize() (Point, Point, bool) {
	if !r.Active || r.Anchor == r.Focus {
		return Point{}, Point{}, false
	}
	if pointLess(r.Focus, r.Anchor) {
		return r.Focus, r.Anchor, true
	}
	return r.Anchor, r.Focus, true
}

func (r Range) SelectedText(lines []string) string {
	start, end, ok := r.Normalize()
	if !ok || len(lines) == 0 {
		return ""
	}
	if end.Line < 0 || start.Line >= len(lines) {
		return ""
	}
	start.Line = clampInt(start.Line, 0, len(lines)-1)
	end.Line = clampInt(end.Line, 0, len(lines)-1)

	parts := make([]string, 0, end.Line-start.Line+1)
	for lineIndex := start.Line; lineIndex <= end.Line; lineIndex++ {
		runes := []rune(lines[lineIndex])
		from := 0
		to := len(runes)
		if lineIndex == start.Line {
			from = clampInt(start.Column, 0, len(runes))
		}
		if lineIndex == end.Line {
			to = clampInt(end.Column, 0, len(runes))
		}
		if from > to {
			return ""
		}
		parts = append(parts, string(runes[from:to]))
	}
	return strings.Join(parts, "\n")
}

func (r Range) CharCount(lines []string) int {
	text := r.SelectedText(lines)
	count := 0
	for _, char := range text {
		if char != '\n' {
			count++
		}
	}
	return count
}

func (r Range) CanCopy(lines []string) bool {
	return r.CharCount(lines) >= minCopyRunes
}

func pointLess(left Point, right Point) bool {
	if left.Line != right.Line {
		return left.Line < right.Line
	}
	return left.Column < right.Column
}

func clampInt(value int, low int, high int) int {
	if high < low {
		return low
	}
	if value < low {
		return low
	}
	if value > high {
		return high
	}
	return value
}

File: crates/roder-tui/src/selection/range.rs
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct Point {
    pub line: usize,
    pub col: usize,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct Range {
    pub anchor: Point,
    pub cursor: Point,
    pub active: bool,
}

impl Range {
    pub fn new(anchor: Point, cursor: Point) -> Self {
        Self {
            anchor,
            cursor,
            active: true,
        }
    }

    pub fn normalized(self) -> Option<(Point, Point)> {
        if !self.active || self.anchor == self.cursor {
            return None;
        }
        if (self.cursor.line, self.cursor.col) < (self.anchor.line, self.anchor.col) {
            Some((self.cursor, self.anchor))
        } else {
            Some((self.anchor, self.cursor))
        }
    }

    pub fn char_count(self, lines: &[impl AsRef<str>]) -> usize {
        selected_text(lines, self)
            .chars()
            .filter(|ch| *ch != '\n')
            .count()
    }

    pub fn is_copyable(self, lines: &[impl AsRef<str>], min_chars: usize) -> bool {
        self.char_count(lines) >= min_chars
    }
}

pub fn selected_text(lines: &[impl AsRef<str>], range: Range) -> String {
    let Some((start, end)) = range.normalized() else {
        return String::new();
    };
    if start.line >= lines.len() {
        return String::new();
    }

    let end_line = end.line.min(lines.len().saturating_sub(1));
    (start.line..=end_line)
        .filter_map(|line_index| {
            let line = lines[line_index].as_ref();
            let len = line.chars().count();
            let from = if line_index == start.line {
                start.col.min(len)
            } else {
                0
            };
            let to = if line_index == end.line {
                end.col.saturating_add(1).min(len)
            } else {
                len
            };
            (from < to).then(|| slice_chars(line, from, to))
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn slice_chars(text: &str, start: usize, end: usize) -> String {
    text.chars()
        .skip(start)
        .take(end.saturating_sub(start))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn range_normalizes_forward_reverse_and_inactive_cases() {
        let forward = Range::new(Point { line: 1, col: 2 }, Point { line: 2, col: 3 });
        assert_eq!(
            forward.normalized(),
            Some((Point { line: 1, col: 2 }, Point { line: 2, col: 3 }))
        );

        let reverse = Range::new(Point { line: 4, col: 9 }, Point { line: 2, col: 1 });
        assert_eq!(
            reverse.normalized(),
            Some((Point { line: 2, col: 1 }, Point { line: 4, col: 9 }))
        );

        assert_eq!(
            Range {
                anchor: Point { line: 1, col: 1 },
                cursor: Point { line: 1, col: 1 },
                active: true,
            }
            .normalized(),
            None
        );
        assert_eq!(
            Range {
                anchor: Point { line: 1, col: 1 },
                cursor: Point { line: 1, col: 3 },
                active: false,
            }
            .normalized(),
            None
        );
    }

    #[test]
    fn selected_text_clips_columns_and_preserves_line_breaks() {
        let lines = ["zero", "alpha", "βeta", "omega"];
        let range = Range::new(Point { line: 1, col: 2 }, Point { line: 2, col: 99 });

        assert_eq!(selected_text(&lines, range), "pha\nβeta");
    }

    #[test]
    fn selected_text_uses_character_indexes_not_bytes() {
        let lines = ["aé日z"];
        let range = Range::new(Point { line: 0, col: 1 }, Point { line: 0, col: 2 });

        assert_eq!(selected_text(&lines, range), "é日");
    }

    #[test]
    fn range_char_count_controls_short_copyability() {
        let lines = ["abcdef"];
        assert!(
            !Range::new(Point { line: 0, col: 1 }, Point { line: 0, col: 2 })
                .is_copyable(&lines, 3)
        );
        assert!(
            Range::new(Point { line: 0, col: 1 }, Point { line: 0, col: 3 }).is_copyable(&lines, 3)
        );
    }
}
